Source/SwiftLintBuiltInRules/Rules/Metrics/CognitiveComplexityRule.swift

```diff
@@ -0,0 +1,266 @@
+                                sum += buffer[index]
+                            } else if sum > 0 {
+                                sum -= buffer[index]
+                            } else if sum < 0 {
+                                sum += buffer[index]
+                            }
+                        }
+                    }
+                }
+                if sum < 0 {
+                    return -sum
+                }
+                return sum
+            }
+            """),
+        ]
+    )
+}
+
+private extension CognitiveComplexityRule {
+    final class Visitor: ViolationsSyntaxVisitor<ConfigurationType> {
+        override func visitPost(_ node: FunctionDeclSyntax) {
+            guard let body = node.body else {
+                return
+            }
+
+            // for legacy reasons, we try to put the violation in the static or class keyword
+            let violationToken = node.modifiers.staticOrClassModifier ?? node.funcKeyword
+            validate(body: body, violationToken: violationToken)
+        }
+
+        override func visitPost(_ node: InitializerDeclSyntax) {
+            guard let body = node.body else {
+                return
+            }
+
+            validate(body: body, violationToken: node.initKeyword)
+        }
+
+        private func validate(body: CodeBlockSyntax, violationToken: TokenSyntax) {
+            let complexity = ComplexityVisitor(
+                ignoresLogicalOperatorSequences: configuration.ignoresLogicalOperatorSequences
+            ).walk(tree: body, handler: \.complexity)
+
+            for parameter in configuration.params where complexity > parameter.value {
+                let reason = "Function should have cognitive complexity \(configuration.length.warning) or less; " +
+                             "currently complexity is \(complexity)"
+
+                let violation = ReasonedRuleViolation(
+                    position: violationToken.positionAfterSkippingLeadingTrivia,
+                    reason: reason,
+                    severity: parameter.severity
+                )
+                violations.append(violation)
+                return
+            }
+        }
+    }
+
+    private class ComplexityVisitor: SyntaxVisitor {
+        private let ignoresLogicalOperatorSequences: Bool
+        private(set) var complexity = 0
+        private var nesting = 0
+
+        init(ignoresLogicalOperatorSequences: Bool) {
+            self.ignoresLogicalOperatorSequences = ignoresLogicalOperatorSequences
+            super.init(viewMode: .sourceAccurate)
+        }
+
+        override func visit(_: ForStmtSyntax) -> SyntaxVisitorContinueKind {
+            nesting += 1
+            return .visitChildren
+        }
+
+        override func visitPost(_: ForStmtSyntax) {
+            nesting -= 1
+            complexity += nesting + 1
+        }
+
+        override func visit(_: IfExprSyntax) -> SyntaxVisitorContinueKind {
+            nesting += 1
```

Comment:
It feels like there's quite a lot of repetition in the methods - it would be nice to abstract that way a bit if possible